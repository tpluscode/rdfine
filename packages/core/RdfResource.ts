/* eslint-disable camelcase,@typescript-eslint/camelcase,no-dupe-class-members */
import RDF from '@rdfjs/data-model'
import { NamespaceBuilder } from '@rdfjs/namespace'
import { NamedNode, DatasetCore, BlankNode, DefaultGraph, Quad_Graph, Term, Literal } from 'rdf-js'
import cf, { SafeClownface, SingleContextClownface } from 'clownface'
import ResourceFactoryImpl, {
  Constructor,
  Mixin,
  ResourceCreationOptions,
  ResourceFactory,
  ResourceIndexer,
} from './lib/ResourceFactory'
import once from 'once'
import TypeCollectionCtor, { TypeCollection } from './lib/TypeCollection'
import { xsd } from '@tpluscode/rdf-ns-builders'

type ObjectOrFactory<T> = T | ((self: RdfResource) => T)

export type ResourceIdentifier = BlankNode | NamedNode
export type ResourceNode<D extends DatasetCore = DatasetCore> = { dataset: D; term: Term; graph?: NamedNode | DefaultGraph }

export interface GetOptions {
  strict: boolean
}

export interface RdfResource<D extends DatasetCore = DatasetCore> {
  readonly id: ResourceIdentifier
  readonly types: TypeCollection<D>
  readonly _selfGraph: SingleContextClownface<ResourceIdentifier, D>
  readonly _unionGraph: SafeClownface<ResourceIdentifier, D>
  readonly _graphId: Quad_Graph
  readonly _parent?: RdfResource<D>
  readonly isAnonymous: boolean
  hasType (type: string | NamedNode): boolean
  /**
   * Gets the value of a property
   * @param property
   */
  get<T extends RdfResource = RdfResource> (property: string | NamedNode): T
  get<T extends RdfResource = RdfResource> (property: string | NamedNode, options?: GetOptions): T | null
  /**
   * Gets the value of a property and ensures that an array will be returned
   * @param property
   */
  getArray<T extends RdfResource = RdfResource> (property: string | NamedNode, options?: GetOptions): T[]
  /**
   * Gets the property value if it's boolean. Throws if it's not
   * @param property
   */
  getBoolean (property: string | NamedNode, options?: GetOptions): boolean
  /**
   * Gets the property value if it's a string. Throws if it's not
   * @param property
   */
  getString (property: string | NamedNode): string
  getString (property: string | NamedNode, options?: GetOptions): string | null
  /**
   * Gets the property value if it's a number. Throws if it's not
   * @param property
   */
  getNumber (property: string | NamedNode): number | null
  getNumber (property: string | NamedNode, options?: GetOptions): number | null
  _getObjects(property: string | NamedNode, options?: GetOptions): SafeClownface<Term, D>
  _create<T extends RdfResource>(term: ResourceNode<D>, mixins?: Mixin[] | [Constructor, ...Mixin[]], options?: ResourceCreationOptions<D>): T & ResourceIndexer
}

export default class RdfResourceImpl<D extends DatasetCore = DatasetCore> implements RdfResource<D> {
  public readonly _selfGraph: SingleContextClownface<ResourceIdentifier, D>
  public readonly _unionGraph: SafeClownface<ResourceIdentifier, D>
  public readonly __initialized: boolean = false
  public readonly _parent?: RdfResource<D>
  private readonly __initializeProperties: (() => boolean)
  public static __ns?: NamespaceBuilder
  public static factory: ResourceFactory = new ResourceFactoryImpl(RdfResourceImpl)
  public static __mixins: Mixin[] = []

  private static _userInitializeProperties(resource: RdfResourceImpl, init: Initializer<RdfResource> = {}): void {
    Object.entries(init)
      .filter(([prop]) => prop !== 'id' && prop !== 'types')
      .forEach(([prop, value]) => {
        (resource as any)[prop] = value
      })

    if (init.types && Array.isArray(init.types)) {
      init.types.forEach(type => resource.types.add(type))
    }
  }

  public constructor(graph: ResourceNode<D>, init: Initializer<any> = {}, parent?: RdfResource<D>) {
    if (graph.term.termType !== 'BlankNode' && graph.term.termType !== 'NamedNode') {
      throw new Error(`RdfResource cannot be initialized from a ${graph.term.termType} node`)
    }

    const selfGraph = cf({
      ...graph,
      term: graph.term,
    })

    if (selfGraph._context[0].graph) {
      this._selfGraph = selfGraph
      this._unionGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term, graph: undefined })
    } else {
      this._selfGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term, graph: RDF.defaultGraph() })
      this._unionGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term })
    }

    this.__initializeProperties = once(() => {
      const self = this as any
      const defaults: Map<string, ObjectOrFactory<any>> = self.constructor.__defaults || new Map<string, ObjectOrFactory<any>>()
      defaults.forEach((value, key) => {
        const currentValue = self[key]
        const valueIsEmptyArray = Array.isArray(currentValue) && currentValue.length === 0
        const valueIsUndefined = typeof currentValue === 'undefined'

        if (!(valueIsEmptyArray || valueIsUndefined)) {
          return
        }

        if (typeof value === 'function') {
          self[key] = value(self)
          return
        }

        self[key] = value
      })

      return true
    })

    this._parent = parent
    this.__initialized = this.__initializeProperties()
    RdfResourceImpl._userInitializeProperties(this, init)
  }

  public get id(): ResourceIdentifier {
    return this._selfGraph.term
  }

  public get _graphId(): Quad_Graph {
    return this._selfGraph._context[0].graph!
  }

  public get types(): TypeCollection<D> {
    return new TypeCollectionCtor(this)
  }

  public get isAnonymous() {
    return this.id.termType === 'BlankNode'
  }

  public hasType(type: string | NamedNode): boolean {
    return this.types.has(type)
  }

  public equals(other: RdfResource | undefined | null): boolean {
    if (!other) {
      return false
    }

    return this.id.equals(other.id)
  }

  public get<T extends RdfResource = RdfResource>(property: string | NamedNode, options?: GetOptions): T | RdfResource<D> | null {
    const objects = this.getArray<T>(property, options)

    if (objects.length > 0) {
      return objects[0]
    }

    return null
  }

  public getArray<T extends RdfResource = RdfResource>(property: string | NamedNode, options?: GetOptions): T[] {
    const values = this._getObjects(property, options)
      .filter(obj => obj.term.termType === 'NamedNode' || obj.term.termType === 'BlankNode')
      .map(obj => {
        return this._create<T>(obj, [], { parent: this })
      })

    if (!values.length) {
      return []
    }

    return values
  }

  public getNumber(property: string | NamedNode): number
  public getNumber(property: string | NamedNode, options?: GetOptions): number | null {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return null
    }

    if (value.term.termType === 'Literal') {
      return parseFloat(value.value)
    }

    throw new Error(`Expected property '${property}' to be a number but found '${value}'`)
  }

  public getString(property: string | NamedNode): string
  public getString(property: string | NamedNode, options?: GetOptions): string | null {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return null
    }

    if (value.term.termType === 'Literal') {
      return value.value
    }

    throw new Error(`Expected property '${property}' to be a literal but found '${value}'`)
  }

  public getBoolean(property: string | NamedNode, options?: GetOptions): boolean {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return false
    }

    if (value.term.termType === 'Literal' && xsd.boolean.equals(value.term.datatype)) {
      return value.term.equals(this._selfGraph.literal(true).term)
    }

    throw new Error(`Expected property '${property}' to be a boolean but found '${value}'`)
  }

  public _getObjects(property: string | NamedNode, { strict }: GetOptions = { strict: false }): SafeClownface<Term, D> {
    const propertyNode = typeof property === 'string' ? this._selfGraph.namedNode(property) : property
    const objects = this._selfGraph.out(propertyNode)

    if (!objects.terms.length && strict) {
      throw new Error(`Value for predicate <${property}> was missing`)
    }

    return objects
  }

  public _create<T extends RdfResource>(term: ResourceNode<D>, mixins?: Mixin[] | [Constructor, ...Mixin[]], options: ResourceCreationOptions<D> = {}): T & ResourceIndexer {
    return (this.constructor as Constructor).factory.createEntity<T>(term, mixins, options)
  }
}

type UserDefinedInterface<T extends RdfResource> = Omit<T, keyof RdfResource>

interface BaseInitializer {
  types?: NamedNode[] | TypeCollection<any>
  id?: RdfResource['id']
}

type InitialNode<Node extends Term = NamedNode | BlankNode> = Node | SingleContextClownface<Node>
type InitialLiteral = InitialNode<Literal>

export type Initializer<T> = Partial<Omit<{
  [P in keyof T]?:
  T[P] extends (infer U)[] ? U extends RdfResource ? Initializer<UserDefinedInterface<U>>[] | InitialNode[]
    : U extends Term ? T[P] | InitialNode<Term>[] : T[P] | InitialLiteral[]
    : T[P] extends RdfResource ? Initializer<UserDefinedInterface<T[P]> & BaseInitializer> | InitialNode :
      T[P] extends Term ? T[P] | InitialNode<T[P]>
        : T[P] | InitialLiteral
}, keyof RdfResource>> & BaseInitializer

type PartialRecursive<T> = T extends object ? { [K in keyof T]?: PartialRecursive<T[K]> } : T

export function fromObject<T extends RdfResource>(initializer: Initializer<UserDefinedInterface<T> & BaseInitializer>): T {
  return initializer as unknown as T
}
