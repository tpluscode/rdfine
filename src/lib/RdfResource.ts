import { Term, NamedNode, DatasetCore, BlankNode } from 'rdf-js'
import cf, { SingleContextClownface } from 'clownface'
import ns from '@rdfjs/namespace'
import once from 'once'

const rdf = ns('http://www.w3.org/1999/02/22-rdf-syntax-ns#')

export interface RdfResource<D extends DatasetCore = DatasetCore> {
  readonly id: Term | null
  readonly types: Term[]
  readonly _node: SingleContextClownface<D>
  hasType (type: string | NamedNode): boolean
}

export default class RdfResourceImpl<D extends DatasetCore = DatasetCore> implements RdfResource<D> {
  public readonly _node: SingleContextClownface<D>
  private readonly __initializeProperties: (() => void)
  public static __ns?: any
  private static __defaults = new Map<string, unknown>()

  public constructor(node: SingleContextClownface<D> | { dataset: D; term: NamedNode | BlankNode; graph?: NamedNode }) {
    const contexts = cf(node).toArray()

    if (contexts.length !== 1) {
      throw new Error('RdfResource can only be initialized from a single node. Got ' + contexts.length)
    }

    this._node = cf(node).toArray()[0]

    this.__initializeProperties = once(() => {
      const self = this as any
      RdfResourceImpl.__defaults.forEach((value, key) => {
        if (typeof self[key] === 'undefined') {
          self[key] = value
        }
      })
    })

    this.__initializeProperties()
  }

  public get id() {
    if (!this._node.term) throw new Error('Graph context does not represent a single node')

    return this._node.term
  }

  public get types() {
    return this._node.out(rdf.type).terms
  }

  public hasType(type: string | NamedNode) {
    const typeNode = typeof type === 'string' ? this._node.namedNode(type) : type

    return this._node.has(rdf.type, typeNode).terms.length > 0
  }
}
