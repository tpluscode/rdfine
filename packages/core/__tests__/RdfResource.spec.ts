import cf, { AnyPointer } from 'clownface'
import $rdf from 'rdf-ext'
import { NamedNode, Term } from 'rdf-js'
import { defaultGraph, namedNode, literal, blankNode } from '@rdf-esm/data-model'
import { skos } from '@tpluscode/rdf-ns-builders'
import RdfResource, { Initializer } from '../RdfResource'
import { parse, ex } from './_helpers'
import { property } from '../index'

describe('RdfResource', () => {
  describe('constructor', () => {
    it('throws when node is literal', () => {
      expect(() => {
        const node = cf({ dataset: $rdf.dataset() }).literal('foo')

        return new RdfResource(node as any)
      }).toThrow()
    })

    it('throws when context represents multiple nodes', () => {
      expect(() => {
        const node = cf({
          dataset: $rdf.dataset(),
          term: [$rdf.blankNode(), $rdf.blankNode()],
        })

        return new RdfResource(node as any)
      }).toThrow()
    })
  })

  describe('equals', () => {
    it('compares id term', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).namedNode('urn:foo:bar')
      const left = new RdfResource(node)
      const right = new RdfResource(node)

      // when
      const areEqual = left.equals(right)

      // then
      expect(areEqual).toBe(true)
    })

    it('compares id with term value', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).namedNode('urn:foo:bar')
      const left = new RdfResource(node)
      const right = $rdf.namedNode('urn:foo:bar')

      // when
      const areEqual = left.equals(right)

      // then
      expect(areEqual).toBe(true)
    })

    it('compares id with graph pointer value', () => {
      // given
      const graph = cf({ dataset: $rdf.dataset() })
      const node = graph.namedNode('urn:foo:bar')
      const left = new RdfResource(node)
      const right = graph.namedNode('urn:foo:bar')

      // when
      const areEqual = left.equals(right)

      // then
      expect(areEqual).toBe(true)
    })

    it('returns false if other resource is falsy', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).namedNode('urn:foo:bar')
      const left = new RdfResource(node)

      // when
      const areEqual = left.equals(null)

      // then
      expect(areEqual).toBe(false)
    })

    it('returns true if "same" blank node is from from same dataset', () => {
      // given
      const dataset = $rdf.dataset()
      const foo = cf({ dataset }).blankNode('foo')
      const bar = cf({ dataset }).blankNode('foo')
      const left = new RdfResource(foo)
      const right = new RdfResource(bar)

      // when
      const areEqual = left.equals(right)

      // then
      expect(areEqual).toBe(true)
    })
  })

  describe('_graphId', () => {
    it('returns the clownface context graph', () => {
      // given
      const node = cf({ dataset: $rdf.dataset(), graph: ex.graph }).blankNode()

      // when
      const res = new RdfResource(node)

      // then
      expect(res._graphId.equals(ex.graph))
    })

    it('is default graph when not initialized explicitly', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()

      // when
      const res = new RdfResource(node)

      // then
      expect(res._graphId.equals(defaultGraph()))
    })
  })

  describe('types', () => {
    it('returns an iterable set of resource\'s types', async () => {
      // given
      const dataset = await parse(`
        @prefix ex: <${ex().value}> .
        
        ex:res a ex:Type1, ex:Type2, ex:Type3, ex:Type4 .
      `)
      const node = cf({
        dataset,
      }).namedNode(ex.res)

      // when
      const tc = new RdfResource(node).types

      // then
      expect([...tc.values()].map(r => r.id)).toEqual(
        expect.arrayContaining([
          ex.Type1, ex.Type2, ex.Type3, ex.Type4,
        ]),
      )
    })
  })

  describe('initializeProperties', () => {
    it('skips id', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      interface Resource extends RdfResource {
        name: string
      }
      class ResourceImpl extends RdfResource implements Resource {
        @property.literal({ path: ex.name })
        name!: string;
      }

      // when
      const resource = new ResourceImpl(node, {
        id: ex.Foo,
        name: 'baz',
      })

      // then
      expect(resource.name).toEqual('baz')
      expect(resource.id.termType).toEqual('BlankNode')
      expect(resource.types.size).toEqual(0)
    })

    it('allows RDF/JS literal to initialize literal properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      interface Resource extends RdfResource {
        name: string
      }
      class ResourceImpl extends RdfResource implements Resource {
        @property.literal({ path: ex.name })
        name!: string;
      }
      const initializer: Initializer<Resource> = {
        name: literal('baz'),
      }

      // when
      const resource = new ResourceImpl(node, initializer)

      // then
      expect(resource.name).toEqual('baz')
    })

    it('allows RDF/JS literal to initialize literal array properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      interface Resource extends RdfResource {
        names: string[]
      }
      class ResourceImpl extends RdfResource implements Resource {
        @property.literal({ path: ex.name, values: 'array' })
        names!: string[];
      }
      const initializer: Initializer<Resource> = {
        names: [literal('bar'), node.literal('baz')],
      }

      // when
      const resource = new ResourceImpl(node, initializer)

      // then
      expect(resource.names).toEqual(
        expect.arrayContaining(['bar', 'baz']),
      )
    })

    it('allows clownface literal to initialize literal properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      interface Resource extends RdfResource {
        name: string
      }
      class ResourceImpl extends RdfResource implements Resource {
        @property.literal({ path: ex.name })
        name!: string;
      }
      const initializer: Initializer<Resource> = {
        name: node.literal('baz'),
      }

      // when
      const resource = new ResourceImpl(node, initializer)

      // then
      expect(resource.name).toEqual('baz')
    })

    it('allows RDF/JS named node to initialize resource properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      interface Resource extends RdfResource {
        other: Resource
      }
      class ResourceImpl extends RdfResource implements Resource {
        @property.resource({ path: ex.other })
        other!: Resource;
      }
      const initializer: Initializer<Resource> = {
        other: namedNode('baz'),
      }

      // when
      const resource = new ResourceImpl(node, initializer)

      // then
      expect(resource.other.id).toEqual(namedNode('baz'))
    })

    it('allows clownface named node to initialize named node properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      interface Resource extends RdfResource {
        other: NamedNode
        others: NamedNode[]
      }
      class ResourceImpl extends RdfResource implements Resource {
        @property({ path: ex.other })
        other!: NamedNode;

        @property({ path: ex.others, values: 'array' })
        others!: NamedNode[];
      }
      const initializer: Initializer<Resource> = {
        other: node.namedNode('bar'),
        others: [node.namedNode('baz')],
      }

      // when
      const resource = new ResourceImpl(node, initializer)

      // then
      expect(resource.other).toEqual(namedNode('bar'))
      expect(resource.others).toEqual(
        expect.arrayContaining([namedNode('baz')]),
      )
    })

    it('allows clownface node to initialize term properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      interface Resource extends RdfResource {
        other: Term
        others: Term[]
      }
      class ResourceImpl extends RdfResource implements Resource {
        @property({ path: ex.other, values: 'array' })
        other!: Term;

        @property({ path: ex.others, values: 'array' })
        others!: Term[];
      }
      const initializer: Initializer<Resource> = {
        other: node.blankNode(),
        others: [node.blankNode(), literal('foo'), namedNode('bar'), node.namedNode('baz')],
      }

      // when
      const resource = new ResourceImpl(node, initializer)

      // then
      expect(resource.others).toEqual(
        expect.arrayContaining([
          expect.objectContaining({ termType: 'BlankNode' }),
          literal('foo'),
          namedNode('bar'),
          namedNode('baz'),
        ]),
      )
    })

    it('allows RDF/JS blank node to initialize resource properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      interface Resource extends RdfResource {
        other: Resource
      }
      class ResourceImpl extends RdfResource implements Resource {
        @property.resource({ path: ex.other })
        other!: Resource;
      }
      const initializer: Initializer<Resource> = {
        other: blankNode(),
      }

      // when
      const resource = new ResourceImpl(node, initializer)

      // then
      expect(resource.other.id.termType).toEqual('BlankNode')
    })

    it('allows RDF/JS nodes to initialize array resource properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      interface Resource extends RdfResource {
        others: Resource[]
      }
      class ResourceImpl extends RdfResource implements Resource {
        @property.resource({ path: ex.other, values: 'array' })
        others!: Resource[];
      }
      const initializer: Initializer<Resource> = {
        others: [blankNode(), namedNode('bar'), node.blankNode(), node.namedNode('baz')],
      }

      // when
      const resource = new ResourceImpl(node, initializer)

      // then
      expect(resource.others.map(o => o.id)).toEqual(
        expect.arrayContaining([
          expect.objectContaining({ termType: 'BlankNode' }),
          namedNode('bar'),
          expect.objectContaining({ termType: 'BlankNode' }),
          namedNode('baz'),
        ]),
      )
    })

    it('add additional types', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      interface Resource extends RdfResource {
        name: string
      }
      class ResourceImpl extends RdfResource implements Resource {
        name!: string;
      }

      // when
      const resource = new ResourceImpl(node, {
        types: [ex.Bar],
      })

      // then
      expect(resource.types.size).toEqual(1)
      expect([...resource.types][0].id.value).toEqual(ex.Bar.value)
    })

    it('allows arbitrary string as RDF property', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()

      // when
      const resource = new RdfResource(node, {
        [skos.prefLabel.value]: 'Foo',
        [skos.broader.value]: {
          [skos.prefLabel.value]: 'Bar',
        },
      })

      // then
      expect(resource.getString(skos.prefLabel)).toEqual('Foo')
      expect(resource.get(skos.broader)!.getString(skos.prefLabel)).toEqual('Bar')
    })

    it('initializes annotated property with nested array of resources which use http properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      class SkosResource extends RdfResource {
        @property.resource({ path: skos.broader, values: 'array' })
        broader!: SkosResource[]
      }

      // when
      const resource = new SkosResource(node, {
        broader: [{
          [skos.prefLabel.value]: 'Bar',
        }],
      })

      // then
      expect(resource.broader[0].getString(skos.prefLabel)).toEqual('Bar')
    })

    it('initializes annotated property with a resource using http properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()
      class SkosResource extends RdfResource {
        @property.resource({ path: skos.broader })
        broader!: SkosResource
      }

      // when
      const resource = new SkosResource(node, {
        broader: {
          [skos.prefLabel.value]: 'Bar',
        },
      })

      // then
      expect(resource.broader.getString(skos.prefLabel)).toEqual('Bar')
    })

    it('initializes nested array resources which use http properties', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()

      // when
      const resource = new RdfResource(node, {
        [skos.broader.value]: [{
          [skos.prefLabel.value]: 'Bar',
        }],
      })

      // then
      expect(resource.get(skos.broader)!.getString(skos.prefLabel)).toEqual('Bar')
    })

    it('assigns id to nested resource assigned to arbitrary URI', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()

      // when
      const resource = new RdfResource(node, {
        [skos.prefLabel.value]: 'Foo',
        [skos.broader.value]: {
          id: ex.Bar,
        },
      })

      // then
      expect(resource.getString(skos.prefLabel)).toEqual('Foo')
      expect(resource.get(skos.broader)!.id.value).toEqual(ex.Bar.value)
    })

    it('assigns named node to arbitrary URI property', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()

      // when
      const resource = new RdfResource(node, {
        [skos.broader.value]: ex.Bar,
      })

      // then
      expect(resource.get(skos.broader)!.id.value).toEqual(ex.Bar.value)
    })

    it('assigns literal node to arbitrary URI property', () => {
      // given
      const node = cf({ dataset: $rdf.dataset() }).blankNode()

      // when
      const resource = new RdfResource(node, {
        [skos.prefLabel.value]: $rdf.literal('Foo', 'en'),
      })

      // then
      expect(resource.pointer.dataset).toMatchSnapshot()
    })
  })

  describe('isAnonymous', () => {
    it('true for blank node', () => {
      // given
      const node = cf({ dataset: $rdf.dataset(), graph: ex.graph }).blankNode()

      // when
      const res = new RdfResource(node)

      // then
      expect(res.isAnonymous).toBe(true)
    })

    it('false for named node', () => {
      // given
      const node = cf({ dataset: $rdf.dataset(), graph: ex.graph }).namedNode(ex.foo)

      // when
      const res = new RdfResource(node)

      // then
      expect(res.isAnonymous).toBe(false)
    })
  })

  describe('getters', () => {
    let graph: AnyPointer

    beforeEach(() => {
      graph = cf({ dataset: $rdf.dataset() })
    })

    describe('get', () => {
      it('returns object resource', () => {
        // given
        const node = graph.blankNode()
        node.addOut(ex.foo, ex.bar)
        const resource = new RdfResource(node)

        // when
        const object = resource.get(ex.foo)

        // then
        expect(object).toBeInstanceOf(RdfResource)
        expect(object?.id.value).toEqual(ex.bar.value)
      })

      it('throws when property has no value and calling strict', () => {
        // given
        const node = graph.blankNode()
        const resource = new RdfResource(node)

        // then
        expect(() => {
          // when
          return resource.get(ex.foo, { strict: true })
        }).toThrow()
      })

      it('returns null when property has no value and calling not strict', () => {
        // given
        const node = graph.blankNode()
        const resource = new RdfResource(node)

        // when
        const object = resource.get(ex.foo, { strict: false })

        // then
        expect(object).toBeNull()
      })
    })

    describe('getArray', () => {
      it('returns empty array when property has no value and calling not strict', () => {
        // given
        const node = graph.blankNode()
        const resource = new RdfResource(node)

        // when
        const object = resource.getArray(ex.foo, { strict: false })

        // then
        expect(object.length).toBe(0)
      })

      it('skips literals', () => {
        // given
        const node = graph.blankNode()
        node.addOut(ex.foo, graph.literal('bar'))
        const resource = new RdfResource(node)

        // when
        const object = resource.getArray(ex.foo, { strict: false })

        // then
        expect(object.length).toBe(0)
      })
    })

    describe('getBoolean', () => {
      it('throws when value is not boolean', () => {
        const node = graph.blankNode()
        node.addOut(ex.foo, ex.bar)
        const resource = new RdfResource(node)

        // then
        expect(() => resource.getBoolean(ex.foo)).toThrow()
      })

      it('return false when value is undefined', () => {
        const node = graph.blankNode()
        const resource = new RdfResource(node)

        // then
        expect(resource.getBoolean(ex.foo)).toBeFalsy()
      })

      it('return the value when it is set', () => {
        const node = graph.blankNode()
        node.addOut(ex.foo, true)
        const resource = new RdfResource(node)

        // then
        expect(resource.getBoolean(ex.foo)).toBeTruthy()
      })
    })

    describe('getNumber', () => {
      it('throws when value is not number', () => {
        const node = graph.blankNode()
        node.addOut(ex.foo, ex.bar)
        const resource = new RdfResource(node)

        // then
        expect(() => resource.getNumber(ex.foo)).toThrow()
      })

      it('return null when value is undefined', () => {
        const node = graph.blankNode()
        const resource = new RdfResource(node)

        // then
        expect(resource.getNumber('foo')).toBeNull()
      })

      it('parses the literal', () => {
        const node = graph.blankNode()
        node.addOut(ex.foo, node.literal(10.5))
        const resource = new RdfResource(node)

        // then
        expect(resource.getNumber(ex.foo)).toBe(10.5)
      })
    })

    describe('getString', () => {
      it('returns string value of literal', () => {
        const node = graph.blankNode()
        node.addOut(ex.foo, 123)
        const resource = new RdfResource(node)

        // then
        expect(resource.getString(ex.foo)).toBe('123')
      })

      it('return null when value is undefined', () => {
        const node = graph.blankNode()
        const resource = new RdfResource(node)

        // then
        expect(resource.getString(ex.foo)).toBeNull()
      })
    })
  })
})
