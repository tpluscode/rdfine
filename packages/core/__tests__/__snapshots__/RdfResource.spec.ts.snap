// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RdfResource initializeProperties assigns literal node to arbitrary URI property: toMatchSnapshot 1`] = `
"_:c14n0 <http://www.w3.org/2004/02/skos/core#prefLabel> \\"Foo\\"@en .
"
`;

exports[`RdfResource initializeProperties initializes T|T[] property with IRI: toMatchSnapshot 1`] = `
"_:c14n0 <http://www.w3.org/2004/02/skos/core#broader> <http://example.com/Foo> .
"
`;
